editor/src/clj/util/http_server.clj

```diff
@@ -295,68 +317,81 @@
-                                           connection)]
+                        (let [{:keys [status headers body]} response]
                           (when-not (integer? status)
                             (throw (ex-info (str "Invalid response status: " status) {:status status})))
                           (when-not (or (nil? headers) (map? headers))
                             (throw (ex-info (str "Invalid response headers: " headers) {:headers headers})))
-                          (reduce-kv #(doto ^Headers %1 (.add %2 %3)) (.getResponseHeaders exchange) headers)
-                          (.sendResponseHeaders
-                            exchange
-                            status
-                            ;; Response length argument:
-                            ;; -1 means empty response body (0 bytes)
-                            ;; 0 is unknown size (chunked)
-                            ;; positive is exact size
-                            (if (nil? connection)
-                              -1
-                              (if-let [explicit-length (get headers "content-length")]
-                                (let [n (Long/parseUnsignedLong explicit-length)]
-                                  (if (zero? n) -1 n))
-                                0)))
-                          (when connection
+                          (let [connection (->connection body)]
                             (try
-                              (with-open [is (io/input-stream connection)]
-                                (try
-                                  (io/copy is (.getResponseBody exchange))
-                                  ;; A browser or remote http client can close
-                                  ;; the connection before we finished sending
-                                  ;; the response: ignore such exceptions since
-                                  ;; it's not an issue.
-                                  (catch IOException _)))
+                              (let [;; Infer content-length and content-type for responses that
+                                    ;; can't do that during response creation. For example, we
+                                    ;; might cache a response with a particular file, but
+                                    ;; the file itself might change between requests
+                                    headers (-> headers
+                                                (provide-header "content-type" (connection-content-type connection))
+                                                (provide-header "content-length" (connection-content-length connection)))
+                                    ;; Maybe don't write the data at all (we still need
+                                    ;; to create the connection to infer content headers)
+                                    connection (if (or (= "HEAD" request-method)
+                                                       (= 304 status))
```

Comment:
We could perhaps use `HttpURLConnection/HTTP_NOT_MODIFIED` or add our own constant for this?